cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.data2.datafabric.dataset;
+
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.module.DatasetDefinitionRegistry;
+import co.cask.cdap.api.dataset.module.DatasetModule;
+import co.cask.cdap.common.lang.ClassLoaders;
+import co.cask.cdap.data2.datafabric.dataset.service.DatasetInstanceService;
+import co.cask.cdap.data2.datafabric.dataset.type.ConstantClassLoaderProvider;
+import co.cask.cdap.data2.datafabric.dataset.type.DatasetClassLoaderProvider;
+import co.cask.cdap.data2.dataset2.DatasetDefinitionRegistryFactory;
+import co.cask.cdap.data2.dataset2.module.lib.DatasetModules;
+import co.cask.cdap.proto.DatasetMeta;
+import co.cask.cdap.proto.DatasetModuleMeta;
+import co.cask.cdap.proto.DatasetTypeMeta;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Objects;
+import com.google.common.base.Throwables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Provides {@link Dataset} instances using {@link DatasetMetaProvider}.
+ * Used by {@link RemoteDatasetFramework} and {@link DatasetInstanceService}.
+ * Use this when you want to control how dataset instances are created. For example, when
+ * you want to obtain a {@link Dataset} instance without having to make remote calls.
+ */
+public class LocalDatasetProvider implements DatasetProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LocalDatasetProvider.class);
+  private final DatasetDefinitionRegistryFactory registryFactory;
+  private final DatasetMetaProvider metaProvider;
+
+  public LocalDatasetProvider(DatasetDefinitionRegistryFactory registryFactory, DatasetMetaProvider metaProvider) {
```

Comment:
Since `LocalDatasetProvider` can only be used by `RemoteDatasetFramework` and `DatasetInstanceService`, let's make it explicit in the constructor by having only the following two constructors - 
```
public LocalDatasetProvider(DatasetDefinitionRegistryFactory registryFactory, RemoteDatasetFramework remoteDatasetFramework)

public LocalDatasetProvider(DatasetDefinitionRegistryFactory registryFactory, DatasetInstanceService datasetInstanceService)

```